### ISSUE TITLE
[BUG?] ESLint doesn't report undefined variable even though it is undefined.

### ISSUE THREAD

**Issue by trusktr** [eslint/eslint#6587]
Is posting that single file enough? Or does ESLint do any cross-file things which would require me to post the repo for reproduction?

**What version of ESLint are you using?**
v2.10.2

**What parser (default, Babel-ESLint, etc.) are you using?**
babel-eslint

**Please show your full configuration:**
```
module.exports = {
    extends: "eslint:recommended",
    installedESLint: true, // ??? what's this for?
    parser: 'babel-eslint',
    parserOptions: {
        ecmaFeatures: {
            experimentalObjectRestSpread: true,
            jsx: true,
            modules: true, // import/export statements
        },
        sourceType: "module"
    },
    env: {
        browser: true,
        commonjs: true,
        es6: true,
        node: true
    },
    plugins: [
        "react",
        "cssx",
    ],
    rules: {
        indent: [ "off", 4, {SwitchCase: 1} ],
        'linebreak-style': [ "error", "unix" ],
        //quotes: [ "error", "single" ],
        //semi: [ "error", "never" ]
        'no-console': 'off',
        'comma-dangle': 'off',
        'no-unused-vars': 'warn',
        'no-constant-condition': 'off',
    }
};
```

**What did you do? Please include the actual source code causing the issue.**
I have some code:

```js
import 'document-register-element'
import makeWebComponentBaseClass from './web-component'

const WebComponent = makeWebComponentBaseClass(window.HTMLElement)

export default
class MotorHTMLBase extends WebComponent {
    createdCallback() {
        super.createdCallback()

        this.node = null // to hold the imperative API Node instance.

        // XXX: "this.mountPromise" vs "this.ready":
        // "ready" seems to be more intuitive on the HTML side because
        // if the user has a reference to a motor-node or a motor-scene
        // and it exists in DOM, then it is already "mounted" from the
        // HTML API perspective. Maybe we can use "mountPromise" for
        // the imperative API, and "ready" for the HTML API. For example:
        //
        // await $('motor-scene')[0].ready // When using the HTML API
        // await node.mountPromise // When using the imperative API
        //
        // Or, maybe we can just use ".ready" in both APIs?...
        this._resolveReadyPromise = null
        this.ready = new Promise(r => this._resolveReadyPromise = r)
    }

    init() {
        this._associateImperativeNode()
    }

    /**
     * This method creates the association between this MotorHTMLNode instance
     * and the imperative Node instance.
     *
     * This method may get called by this.init, but can also be called by
     * the Node class if Node is used imperatively. See Node#constructor.
     *
     * @private
     *
     * @param {Node} imperativeMotorNode The Node to associate with this
     * MotorHTMLNode. This parameter is only used in Node#constructor, and this
     * happens when using the imperative form infamous instead of the HTML
     * interface of infamous. When the HTML interface is used, this gets called
     * first without an imperativeMotorNode argument and the call to this in
     * Node#constructor will then be a noop. Basically, either this gets called
     * first by MotorHTMLNode, or first by Node, depending on which API is used
     * first.
     */
    _associateImperativeNode(imperativeMotorNode) {
        // console.log(' -- associating imperative node')
        debugger
        console.log(' - imperative node?', imperativeMotorNode)
        if (!this.node) {
            if (imperativeMotorNode && imperativeMotorNode instanceof Node) {
                debugger
                this.node = imperativeMotorNode
            }
            else {
                debugger
                this.node = this._makeImperativeNode()
            }

            this._signalWhenReady()
        }
    }

    // This method should be overriden by child classes. It should return the
    // imperative-side instance that the HTML-side class (this) corresponds to.
    _makeImperativeNode() {
        throw new TypeError('This method should be implemented by class extening MotorHTMLBase.')
    }

    async _signalWhenReady() {
        await this.node.mountPromise
        this._resolveReadyPromise()
    }
}
```

**What did you expect to happen?**

The file does not define the `Node` variable on line 55 (it should import it, but I forgot), yet ESLint is not reporting an undefined variable error like it normally does.

The thing with my code is that I caused an infinite recursion due to that variable being `undefined`, which made me go on a debug frenzy for a while until I noticed I forgot to import `Node`. That would've been helpful!

**What actually happened? Please include the actual, raw output from ESLint.**
Command line output (am I doing it right? I am using Atom's linting package, not command line):

```
infamous+infamous git:master ❯ eslint -c ~/.eslintrc.js src/motor-html/base.js
```

**Comment by trusktr** [comment]
The thing with my code is that I caused an infinite recursion due to that variable being `undefined`, which made me go on a debug frenzy for a while until I noticed I forgot to import `Node`. That would've been helpful!

**Comment by trusktr** [comment]
Is posting that single file enough? Or does ESLint do any cross-file things which would require me to post the repo for reproduction?

**What version of ESLint are you using?**
v2.10.2

**What parser (default, Babel-ESLint, etc.) are you using?**
babel-eslint

**Please show your full configuration:**
```
module.exports = {
    extends: "eslint:recommended",
    installedESLint: true, // ??? what's this for?
    parser: 'babel-eslint',
    parserOptions: {
        ecmaFeatures: {
            experimentalObjectRestSpread: true,
            jsx: true,
            modules: true, // import/export statements
        },
        sourceType: "module"
    },
    env: {
        browser: true,
        commonjs: true,
        es6: true,
        node: true
    },
    plugins: [
        "react",
        "cssx",
    ],
    rules: {
        indent: [ "off", 4, {SwitchCase: 1} ],
        'linebreak-style': [ "error", "unix" ],
        //quotes: [ "error", "single" ],
        //semi: [ "error", "never" ]
        'no-console': 'off',
        'comma-dangle': 'off',
        'no-unused-vars': 'warn',
        'no-constant-condition': 'off',
    }
};
```

**What did you do? Please include the actual source code causing the issue.**
^ above

**What did you expect to happen?**
^ above

**What actually happened? Please include the actual, raw output from ESLint.**
Command line output (am I doing it right? I am using Atom's linting package, not command line):

```
infamous+infamous git:master ❯ eslint -c ~/.eslintrc.js src/motor-html/base.js
```

**Comment by platinumazure** [comment]
Hi @trusktr, unfortunately the rule is behaving as expected. By including the "browser" environment in your config file, any global variables that can be expected in a browser environment are also implicitly added as known globals, meaning no-undef (which you are using by extending `eslint:recommended`) will not flag them.

If this is a script that is not actually meant to be used on the browser, I would suggest removing `"browser": true` from your environments (or replacing it with `"browser": false` to be explicit).

Let me know if this doesn't make sense.

**Comment by trusktr** [comment]
Ah, thanks for letting me know! I guess Ican make a local eslintrc, instead of using my global one.
